app/controllers/saml_idp_controller.rb

```diff
@@ -45,6 +45,91 @@ def disable_caching
     response.headers['Pragma'] = 'no-cache'
   end
 
+  def store_sp_data
+    session[:sp_data] = {
+      provider: saml_request.service_provider.identifier,
+      authn_context: requested_authn_context
+    }
+
+    return if relay_state_params.blank?
+
+    session[:sp_data].merge!(ial_token: ial_token)
+  end
+
+  def valid_authn_contexts
+    [Saml::Idp::Constants::LOA1_AUTHNCONTEXT_CLASSREF,
+     Saml::Idp::Constants::LOA3_AUTHNCONTEXT_CLASSREF]
+  end
+
+  def process_invalid_authn_context
+    logger.info "Invalid authn context #{requested_authn_context} requested"
+    render nothing: true, status: :bad_request
+  end
+
+  def requested_authn_context
+    if saml_request.requested_authn_context
+      saml_request.requested_authn_context
+    else
+      logger.info 'authn_context is missing'
+      nil
+    end
+  end
+
+  def check_ial_token
+    return if current_user.identity_verified?
+    route_idv_request if ial_token
+  end
+
+  def route_idv_request
+    store_saml_request_in_session
+    link_identity_from_session_data
+    redirect_to idp_index_url
+  end
+
+  # stores original SAMLRequest in session to continue SAML Authn flow
+  def store_saml_request_in_session
+    session[:SAMLRequest] = request.original_url
+  end
+
+  def relay_state_params
+    JSON.parse(URI.unescape(params.fetch('RelayState', '{}')))
+  end
+
+  def ial_token
+    relay_state_params['token'] || relay_state_params['ial_token']
+  end
+
+  def validate_saml_logout_param
+    prepare_saml_logout_response if params[:SAMLResponse].present?
+    prepare_saml_logout_request if params[:SAMLRequest].present?
+  end
+
+  def prepare_saml_logout_response
+    @saml_response = OneLogin::RubySaml::Logoutresponse.new(params[:SAMLResponse])
+  end
+
+  def prepare_saml_logout_request
+    validate_saml_request
+    return if session[:logout_response]
+    # store originating SP's logout response in the user session
+    # for final step in SLO
+    session[:logout_response] = logout_response_builder.signed
+    session[:logout_response_url] = saml_request.response_url
+  end
+
+  def saml_response
+    encode_response(
+      current_user,
+      authn_context_classref: requested_authn_context,
+      reference_id: auth_identity.session_uuid.gsub(/^_/, ''),
+      encryption: current_service_provider.encryption_opts
+    )
+  end
+
+  def current_service_provider
+    ServiceProvider.new(saml_request.service_provider.identifier)
+  end
+
```

Comment:
Please move these to `saml_idp_auth_concern.rb` and replace the `saml_response` method that might be there.